src/lib.rs

```diff
@@ -629,6 +635,13 @@ impl fmt::Debug for Connection {
 
 struct InnerConnection {
     db: *mut ffi::sqlite3,
+    // It's unsafe to call `sqlite3_close` while another thread is performing
+    // a `sqlite3_interrupt`, and vice versa, so we take this mutex during
+    // those functions. This protects a copy of the `db` pointer (which is
+    // cleared on closing), however the main copy, `db`, is unprotected.
+    // Otherwise, a long running query would prevent calling interrupt, as
+    // interrupt would only acquire the lock after the query's completion.
+    interrupt_lock: Arc<Mutex<*mut ffi::sqlite3>>,
```

Comment:
RwLock ?
Many readers (InterruptHandle) but only one writer (during close).

Reply:
SQLite documentation does not specify if it is safe to call `sqlite3_interrupt` multiple times/concurrently...
So I guess a Mutex is ok.